Return requested content type error response in case of URI Violation and not always text/html

**Jetty version(s)**
Jetty 12.0.7

**Jetty Environment**
core

**Java version/vendor** `(use: java -version)`
openjdk version "21.0.2"

**Description**

When a client requests a path with an URI violation, the server currently always returns a text/html error response regardless of what the client specified in the `Accept` header, even though the `ErrorHandler` class checks for that header.

When debugging,. it seems that at that point in the ErrorHandler the request headers are completely empty, maybe unparsed?


**How to reproduce?**

The following test requests a JSON response content type but fails with

```
[ERROR]   ServerTest.testGETWithURIViolation:145 
Expected: is "application/json;charset=utf-8"
     but: was "text/html;charset=ISO-8859-1"
```

```java
public class ServerTest
{
   // A test case added to `jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/ServerTest.java`
   
    public void testGETWithURIViolation() throws Exception
    {
        _context.setHandler(new Handler.Abstract()
        {
            @Override
            public boolean handle(Request request, Response response, Callback callback)
            {
                response.getHeaders().put(HttpHeader.CONTENT_TYPE, "text/plain");
                Content.Sink.write(response, true, "Hello", callback);
                return true;
            }
        });
        _server.start();

        String request = """
                GET /path/with//uri/violation HTTP/1.0
                Host: hostname
                Accept: application/json
                
                """;
        HttpTester.Response response = HttpTester.parseResponse(_connector.getResponse(request));

        assertThat(response.getStatus(), is(HttpStatus.BAD_REQUEST_400));
        assertThat(response.getField(HttpHeader.CONTENT_TYPE).getValue(), is("application/json"));
    }

    // ...
}

```

Expected returned content
```javascript
{
"cause0":"org.eclipse.jetty.http.BadMessageException: 400: Ambiguous URI empty segment",
"message":"Ambiguous URI empty segment",
"url":"/path/with//uri/violation",
"status":"400"
}

```

But was following
```html
<html>
<head>
<meta http-equiv="Content-Type" content="text/html;charset=ISO-8859-1"/>
<title>Error 400 Ambiguous URI empty segment</title>
</head>
<body>
<h2>HTTP ERROR 400 Ambiguous URI empty segment</h2>
<table>
<tr><th>URI:</th><td>/badURI</td></tr>
<tr><th>STATUS:</th><td>400</td></tr>
<tr><th>MESSAGE:</th><td>Ambiguous URI empty segment</td></tr>
<tr><th>CAUSED BY:</th><td>org.eclipse.jetty.http.BadMessageException: 400: Ambiguous URI empty segment</td></tr>
</table>
<h3>Caused by:</h3><pre>org.eclipse.jetty.http.BadMessageException: 400: Ambiguous URI empty segment
        at org.eclipse.jetty.server@12.0.8-SNAPSHOT/org.eclipse.jetty.server.internal.HttpConnection$HttpStreamOverHTTP1.headerComplete(HttpConnection.java:1187)
        at org.eclipse.jetty.server@12.0.8-SNAPSHOT/org.eclipse.jetty.server.internal.HttpConnection$RequestHandler.headerComplete(HttpConnection.java:953)
        at org.eclipse.jetty.http@12.0.8-SNAPSHOT/org.eclipse.jetty.http.HttpParser.parseFields(HttpParser.java:1344)
        at org.eclipse.jetty.http@12.0.8-SNAPSHOT/org.eclipse.jetty.http.HttpParser.parseNext(HttpParser.java:1637)
        at org.eclipse.jetty.server@12.0.8-SNAPSHOT/org.eclipse.jetty.server.internal.HttpConnection.parseRequestBuffer(HttpConnection.java:568)
        at org.eclipse.jetty.server@12.0.8-SNAPSHOT/org.eclipse.jetty.server.internal.HttpConnection.onFillable(HttpConnection.java:391)
        at org.eclipse.jetty.io@12.0.8-SNAPSHOT/org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:322)
        at org.eclipse.jetty.io@12.0.8-SNAPSHOT/org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:99)
        at org.eclipse.jetty.io@12.0.8-SNAPSHOT/org.eclipse.jetty.io.ByteArrayEndPoint.lambda$new$0(ByteArrayEndPoint.java:58)
        at org.eclipse.jetty.util@12.0.8-SNAPSHOT/org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:971)
        at org.eclipse.jetty.util@12.0.8-SNAPSHOT/org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.doRunJob(QueuedThreadPool.java:1201)
        at org.eclipse.jetty.util@12.0.8-SNAPSHOT/org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1156)
        at java.base/java.lang.Thread.run(Thread.java:1583)
</pre>
<hr/><a href="https://eclipse.org/jetty">Powered by Jetty:// 12.0.8-SNAPSHOT</a><hr/>

</body>
</html>

```

--- a/jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ErrorHandler.java
+++ b/jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ErrorHandler.java
@@ -127,7 +127,7 @@ public class ErrorHandler implements Request.Handler
                 callback.succeeded();
                 return;
             }
-            acceptable = Collections.singletonList(Type.TEXT_HTML.asString());
+            acceptable = Collections.singletonList(defaultResponseMimeType().asString());
         }
         List<Charset> charsets = request.getHeaders().getQualityCSV(HttpHeader.ACCEPT_CHARSET).stream()
             .map(s ->
@@ -161,6 +161,10 @@ public class ErrorHandler implements Request.Handler
                 return;
         }
         callback.succeeded();
+    }
+
+    protected Type defaultResponseMimeType() {
+        return Type.TEXT_HTML;
     }
 
     protected boolean generateAcceptableResponse(Request request, Response response, Callback callback, String contentType, List<Charset> charsets, int code, String message, Throwable cause) throws IOException

--- a/jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ErrorHandler.java
+++ b/jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ErrorHandler.java
@@ -74,7 +74,7 @@ public class ErrorHandler implements Request.Handler
 
     boolean _showStacks = true;
     boolean _showMessageInTitle = true;
-    Type _defaultResponseMimeType = Type.TEXT_HTML;
+    String _defaultResponseMimeType = Type.TEXT_HTML.asString();
     HttpField _cacheControl = new PreEncodedHttpField(HttpHeader.CACHE_CONTROL, "must-revalidate,no-cache,no-store");
 
     public ErrorHandler()
@@ -128,7 +128,7 @@ public class ErrorHandler implements Request.Handler
                 callback.succeeded();
                 return;
             }
-            acceptable = Collections.singletonList(_defaultResponseMimeType.asString());
+            acceptable = Collections.singletonList(_defaultResponseMimeType);
         }
         List<Charset> charsets = request.getHeaders().getQualityCSV(HttpHeader.ACCEPT_CHARSET).stream()
             .map(s ->
@@ -482,7 +482,7 @@ public class ErrorHandler implements Request.Handler
      * @return The mime type to be used when a client does not specify an Accept header, or the request did not fully parse
      */
     @ManagedAttribute("Mime type to be used when a client does not specify an Accept header, or the request did not fully parse")
-    public Type getDefaultResponseMimeType()
+    public String getDefaultResponseMimeType()
     {
         return _defaultResponseMimeType;
     }
@@ -490,9 +490,9 @@ public class ErrorHandler implements Request.Handler
     /**
      * @param defaultResponseMimeType The mime type to be used when a client does not specify an Accept header, or the request did not fully parse
      */
-    public void setDefaultResponseMimeType(Type defaultResponseMimeType)
+    public void setDefaultResponseMimeType(String defaultResponseMimeType)
     {
-        _defaultResponseMimeType = defaultResponseMimeType;
+        _defaultResponseMimeType = Objects.requireNonNull(defaultResponseMimeType);;
     }
 
     protected void write(Writer writer, String string) throws IOException

--- a/jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ErrorHandler.java
+++ b/jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ErrorHandler.java
@@ -74,6 +74,7 @@ public class ErrorHandler implements Request.Handler
 
     boolean _showStacks = true;
     boolean _showMessageInTitle = true;
+    Type _defaultResponseMimeType = Type.TEXT_HTML;
     HttpField _cacheControl = new PreEncodedHttpField(HttpHeader.CACHE_CONTROL, "must-revalidate,no-cache,no-store");
 
     public ErrorHandler()
@@ -127,7 +128,7 @@ public class ErrorHandler implements Request.Handler
                 callback.succeeded();
                 return;
             }
-            acceptable = Collections.singletonList(defaultResponseMimeType().asString());
+            acceptable = Collections.singletonList(_defaultResponseMimeType.asString());
         }
         List<Charset> charsets = request.getHeaders().getQualityCSV(HttpHeader.ACCEPT_CHARSET).stream()
             .map(s ->
@@ -161,10 +162,6 @@ public class ErrorHandler implements Request.Handler
                 return;
         }
         callback.succeeded();
-    }
-
-    protected Type defaultResponseMimeType() {
-        return Type.TEXT_HTML;
     }
 
     protected boolean generateAcceptableResponse(Request request, Response response, Callback callback, String contentType, List<Charset> charsets, int code, String message, Throwable cause) throws IOException
@@ -479,6 +476,23 @@ public class ErrorHandler implements Request.Handler
     public void setShowMessageInTitle(boolean showMessageInTitle)
     {
         _showMessageInTitle = showMessageInTitle;
+    }
+
+    /**
+     * @return The mime type to be used when a client does not specify an Accept header, or the request did not fully parse
+     */
+    @ManagedAttribute("Mime type to be used when a client does not specify an Accept header, or the request did not fully parse")
+    public Type getDefaultResponseMimeType()
+    {
+        return _defaultResponseMimeType;
+    }
+
+    /**
+     * @param defaultResponseMimeType The mime type to be used when a client does not specify an Accept header, or the request did not fully parse
+     */
+    public void setDefaultResponseMimeType(Type defaultResponseMimeType)
+    {
+        _defaultResponseMimeType = defaultResponseMimeType;
     }
 
     protected void write(Writer writer, String string) throws IOException

--- a/jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/ErrorHandlerTest.java
+++ b/jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/ErrorHandlerTest.java
@@ -167,10 +167,11 @@ public class ErrorHandlerTest
         errorHandler.setDefaultResponseMimeType(MimeTypes.Type.APPLICATION_JSON.asString());
         server.setErrorHandler(errorHandler);
 
-        String rawResponse = connector.getResponse(
-                "GET / HTTP/1.1\r\n" +
-                        "Host: Localhost\r\n" +
-                        "\r\n");
+        String rawResponse = connector.getResponse("""
+                        GET / HTTP/1.1
+                        Host: Localhost
+                        
+                        """);
 
         HttpTester.Response response = HttpTester.parseResponse(rawResponse);
 

--- a/jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/ErrorHandlerTest.java
+++ b/jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/ErrorHandlerTest.java
@@ -164,7 +164,7 @@ public class ErrorHandlerTest
     public void test404NoAcceptButSpecifiedDefaultResponseMimeType() throws Exception
     {
         ErrorHandler errorHandler = new ErrorHandler();
-        errorHandler.setDefaultResponseMimeType(MimeTypes.Type.APPLICATION_JSON);
+        errorHandler.setDefaultResponseMimeType(MimeTypes.Type.APPLICATION_JSON.asString());
         server.setErrorHandler(errorHandler);
 
         String rawResponse = connector.getResponse(

--- a/jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/ErrorHandlerTest.java
+++ b/jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/ErrorHandlerTest.java
@@ -26,6 +26,7 @@ import org.eclipse.jetty.http.BadMessageException;
 import org.eclipse.jetty.http.HttpHeader;
 import org.eclipse.jetty.http.HttpStatus;
 import org.eclipse.jetty.http.HttpTester;
+import org.eclipse.jetty.http.MimeTypes;
 import org.eclipse.jetty.io.Content;
 import org.eclipse.jetty.io.QuietException;
 import org.eclipse.jetty.logging.StacklessLogging;
@@ -155,6 +156,30 @@ public class ErrorHandlerTest
         assertThat("Response Content-Type", response.get(HttpHeader.CONTENT_TYPE), containsString("text/html;charset=ISO-8859-1"));
         assertThat(response.get(HttpHeader.DATE), notNullValue());
         assertThat(response.getContent(), containsString("content=\"text/html;charset=ISO-8859-1\""));
+
+        assertContent(response);
+    }
+
+    @Test
+    public void test404NoAcceptButSpecifiedDefaultResponseMimeType() throws Exception
+    {
+        ErrorHandler errorHandler = new ErrorHandler();
+        errorHandler.setDefaultResponseMimeType(MimeTypes.Type.APPLICATION_JSON);
+        server.setErrorHandler(errorHandler);
+
+        String rawResponse = connector.getResponse(
+                "GET / HTTP/1.1\r\n" +
+                        "Host: Localhost\r\n" +
+                        "\r\n");
+
+        HttpTester.Response response = HttpTester.parseResponse(rawResponse);
+
+        assertThat("Response status code", response.getStatus(), is(404));
+        assertThat("Response Content-Length", response.getField(HttpHeader.CONTENT_LENGTH).getIntValue(), greaterThan(0));
+        assertThat("Response Content-Type", response.get(HttpHeader.CONTENT_TYPE), containsString("application/json"));
+        assertThat(response.get(HttpHeader.DATE), notNullValue());
+        assertThat(response.getContent(), containsString("\"status\":\"404\""));
+        assertThat(response.getContent(), containsString("\"message\":\"Not Found\""));
 
         assertContent(response);
     }

